File: lec-2/convServer.py
#!/usr/bin/env python

#******************************************************************************
#
#  CS 6421 - Simple Conversation
#  Execution:    python convServer.py portnum
#  Team member:  Yuan Gao, Bei Xia
#  Conversion:   Pounds <---> Ounces
#  Default Port: 1111
#******************************************************************************

import socket
import sys

## Function to process requests
def process(conn):
    conn.send("**************************************************************************\n")
    conn.send("Welcome to the Pounds (lbs) <---> Ounces (oz) Python-based conversion server!\n")
    conn.send("Conversion: <input unit> <output unit> <input amount>(lbs oz XX)or(oz lbs XX)\n")
    conn.send("**************************************************************************\n")
    # read userInput from client
    userInput = conn.recv(BUFFER_SIZE)
    if not userInput:
        print "Error reading message"
        sys.exit(1)

    print "Received message: ", userInput
    # TODO: add convertion function here, reply = func(userInput)
    try:
	#split the string, input, output, amount
    	strinput = userInput.split(" ")
	#pounds converse to ounces
	if strinput[0]==("lbs") and strinput[1]==("oz"):
		conn.send(str(float(strinput[2])*16)+"\n")
	#ounces converse to pounds
	elif strinput[0]==("oz") and strinput[1]==("lbs"):
	        conn.send(str(float(strinput[2])/16)+"\n")
	else:
	        conn.send("**************************************************************************\n")
        	conn.send("Invalid Input\n")
	        conn.send("Conversion: <input unit> <output unit> <input amount>(lbs oz XX)or(oz lbs XX)\n")
        	conn.send("**************************************************************************\n")
    except:	           
  	conn.send("**************************************************************************\n")          
        conn.send("Invalid Input\n")                  
  	conn.send("Conversion: <input unit> <output unit> <input amount>(lbs oz XX)or(oz lbs XX)\n")
    	conn.send("**************************************************************************\n")
    conn.close()
### Main code run when program is started
BUFFER_SIZE = 1024
interface = ""

# if input arguments are wrong, print out usage
#if len(sys.argv) != 2:
#    print >> sys.stderr, "usage: python {0} portnum\n".format(sys.argv[0])
#    sys.exit(1)

#portnum = int(sys.argv[1])
portnum = 1111
# create socket
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.bind((interface, portnum))
s.listen(5)
print 'Started Server on port 1111'
# accept connection and print out info of client
while True:
	conn, addr = s.accept()
	print 'Accepted connection from client', addr
	process(conn)	
s.close()

File: lec-2/ConvServer.java
/******************************************************************************
 *
 *  CS 6421 - Simple Conversation
 *  Compilation:  javac ConvServer.java
 *  Execution:    java ConvServer port
 *  Team Member:  Yuan Gao, Bei Xia
 *  Conversion:   Dollars <---> Ounces of bananas
 *  Default Port: 2222
 *  % java ConvServer portnum
 ******************************************************************************/

import java.net.Socket;
import java.net.ServerSocket;
import java.io.IOException;
import java.io.PrintWriter;
import java.net.UnknownHostException;
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class ConvServer {

    public static void process (Socket clientSocket) throws IOException {
        // open up IO streams
        BufferedReader in = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
        PrintWriter out = new PrintWriter(clientSocket.getOutputStream(), true);

        /* Write a welcome message to the client */
	out.println("**************************************************************************");
	out.println("Welcome to the Dollars ($) <---> Ounces (oz) Java-based conversion server!");
	out.println("Conversion: <input unit> <output unit> <input amount>($ oz XX)or(oz $ XX)");
	out.println("**************************************************************************");
        /* read and print the client's request */
        // readLine() blocks until the server receives a new line from client
        String userInput;
        if ((userInput = in.readLine()) == null) {
            System.out.println("Error reading message");
            out.close();
            in.close();
            clientSocket.close();
        }

        System.out.println("Received message: " + userInput);
        //--TODO: add your converting functions here, msg = func(userInput);
	try{
	// split the string, input, output, amount
	String str[] = userInput.split(" ");
	//dollars converse to ounces
	if(str[0].equals("$")&&str[1].equals("oz"))
		out.println(Float.parseFloat(str[2])*20);
	//ounces converse to dollars
	else if(str[0].equals("oz")&&str[1].equals("$"))
		out.println(Float.parseFloat(str[2])/20);
        else{
		out.println("**************************************************************************");
		out.println("Invalid Input");
		out.println("Conversion: <input unit> <output unit> <input amount>($ oz XX)or(oz $ XX)");
		out.println("**************************************************************************");
	}}
	catch(Exception e){
		out.println("**************************************************************************");
		out.println("Invalid Input");
		out.println("Conversion: <input unit> <output unit> <input amount>($ oz XX)or(oz $ XX)");
		out.println("**************************************************************************");
	}
	// close IO streams, then socket
        out.close();
        in.close();
        clientSocket.close();
    }

    public static void main(String[] args) throws Exception {

        //check if argument length is invalid
        //if(args.length != 1) {
        //    System.err.println("Usage: java ConvServer port");
        //}
        // create socket
        //int port = Integer.parseInt(args[0]);
        int port = 2222;
	ServerSocket serverSocket = new ServerSocket(port);
        System.err.println("Started server on port " + port);

        // wait for connections, and process
        try {
            while(true) {
                // a "blocking" call which waits until a connection is requested
                Socket clientSocket = serverSocket.accept();
                System.err.println("\nAccepted connection from client");
                process(clientSocket);
            }

        }catch (IOException e) {
            System.err.println("Connection Error");
        }
        System.exit(0);
    }
}
